### PR TITLE
Elm-analyse code actions

### PR DESCRIPTION
**Pull request by antew**
I think this is about ready now, it adds in a code action provider that supports elm-analyse.  
The upstream set of changes for `elm-analyse` is https://github.com/stil4m/elm-analyse/pull/209.

fixer

Features:
- `elm-analyse` messages that can be fixed now show up as code actions in the editor!
- `elm-format` is run on the fixed text

Changes:
- Moved common diffing code into `src/util/diff.ts`, it's now used by the `documentFormattingProvder` and `elmAnalyseDiagnostics` to generate changelists for the document so I moved it over there and have both importing it.
- Moved `elmAnalyseDiagnostics` up a level in the code so that it can be given to the `diagnosticsProvider` and the `codeActionProvider`.
- Broke out a block of code in the `documentFormattingProvider` into a public function for formatting some text, this way elm-analyse can make use of `elm-format` as well and we can apply the already formatted fix to the document.

Misc:
- I noticed that the `TextDocumentEvents` was importing `EventEmitter` from `ws`, so I switched it to use it from `events` like the other files.

### CODE REVIEW

File: src/providers/diagnostics/elmAnalyseDiagnostics.ts [hunk 1]
```diff
@@ -1,32 +1,68 @@
...
+  "UnusedImportedVariable",
+  "UnusedImportAlias",
+  "UnusedPatternVariable",
+  "UnusedTypeAlias",
+  "MultiLineRecordFormatting",
+  "DropConsOfItemAndList",
+  "DuplicateImport",
+];
+const ELM_ANALYSE = "elm-analyse";
+export const CODE_ACTION_ELM_ANALYSE = "elm-analyse-fixer";
```
Comment (Razzeee) [inline src/providers/diagnostics/elmAnalyseDiagnostics.ts]: As you seem to be using this for the command name, I think we should try to mimic the build in command names like these https://code.visualstudio.com/api/references/commands

**Comment by Razzeee** [comment]
Thank you. Code looks good, just a minor.

Will also need a new elm-analyse version.

File: src/providers/diagnostics/elmAnalyseDiagnostics.ts (hunk 1 above)
Comment (antew) [reply src/providers/diagnostics/elmAnalyseDiagnostics.ts]: Sounds good! I changed it over to `elmLanguageServer.elmAnalyseFixer` with 38e7490, if you thing something else work work better let me know.

**Comment by antew** [comment]
I pushed a branch with the built `elm-analyse` files so that it is easier to test out, it's available iwth `npm install "antew/elm-analyse#lsp-fixer-built" --save`

File: src/providers/diagnostics/elmAnalyseDiagnostics.ts (hunk 1 above)
Comment (Razzeee) [reply src/providers/diagnostics/elmAnalyseDiagnostics.ts]: I think we settled for `elmLS` instead of `elmLanguageServer`

Comment (antew) [reply src/providers/diagnostics/elmAnalyseDiagnostics.ts]: Sweet, updated w/398ef5f

**Pull request merged**